### PR TITLE
Deep Quantile Transformer

### PR DESCRIPTION
**Pull request by djgagne**
I just added a Deep Quantile Transformer class that will compute the quantile transform on n-dimensional image-like data. It behaves very similarly to the sklearn quantile transformer, unless you turn on stochastic mode, which will randomly pick a point between the left and right interpolation of the quantile values instead of the mid point. 

I also implemented a distributed standard scaler that you can run separately on each data file and then aggregate in map-reduce fashion. Doing a min-max scaler in similar fashion is pretty straightforward. A distributed quantile transform is much less straightforward, and I have been struggling with coming up with a way to do it. There is an algorithm called the t-digest that can do it [blog post](https://dataorigami.net/2015/03/19/Percentile-and-Quantile-Estimation-of-Big-Data-The-t-Digest.html), but the Python package for it has a dependency that errors when I try to install it with Python 3.8. Other people complained in the issues but no resolution has been made. We would either have to implement our own version of this algorithm or will need to wait on distributed quantile transforms. I should note that scikit-learn by default downsamples the QuantileTransformer to 10000 examples, which is quite problematic if you are trying to apply it to large datasets.

### CODE REVIEW

File: bridgescaler/distributed.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import numpy as np
+from copy import copy
+
+
+class DStandardScaler(object):
+    """
```
Comment (charlie-becker) [inline bridgescaler/distributed.py:6]: Can you add a test for the distributed scaler?

File: bridgescaler/tests/deep_test.py [hunk 2]
```diff
@@ -51,4 +53,34 @@ def test_deep_minmax_scaler():
...
+    reg_qs = QuantileTransformer(n_quantiles=1000, subsample=dim * dim * n_ex)
+    def flatten_to_2D(X):
+        return np.reshape(X, newshape=(X.shape[0] * X.shape[1] * X.shape[2], X.shape[-1]))
+
+    x_flat = flatten_to_2D(x)
+    x_scaled = reg_qs.fit_transform(x_flat)
+    x_tel_2 = np.reshape(reg_qs.inverse_transform(x_scaled), newshape=(x.shape[0], x.shape[1], x.shape[2], x.shape[3]))
+    full_diff = np.abs(x - x_telephone).ravel()
+    reg_diff = np.abs(x_tel_2 - x).ravel()
+    assert x_transformed.shape == x.shape, "Shape mismatch"
```
Comment (charlie-becker) [inline bridgescaler/tests/deep_test.py:82]: Can you add a similar assert for `full_diff == reg_diff`? I realize these won't match exactly, but we should test that they are similar.

**Review by charlie-becker — CHANGES_REQUESTED** [review]
This looks good, a couple of minor test requests in comments.

File: bridgescaler/distributed.py (hunk 1 above)
Comment (djgagne) [reply bridgescaler/distributed.py:6]: Done! Found a bunch of issues in the process that I've fixed.

File: bridgescaler/tests/deep_test.py (hunk 2 above)
Comment (djgagne) [reply bridgescaler/tests/deep_test.py:82]: I added a test for this and it passes.

**Comment by djgagne** [comment]
@charlie-becker I should have addressed all the issues you brought up and made a few more changes for added functionality.

**Review by charlie-becker — APPROVED** [review]

**Comment by charlie-becker** [comment]
Thanks for adding the tests! Glad it helped identify some issues. I have tested and visualized the code on a sample batch of project data (with nans) and appears to be functioning as intended. Looks good. Merging.

**Pull request merged**